Adding a helper to merge two colors with different opacities.

mohammadcazig opened the pull request:
Having Multiple layers with different opacity in our components can cause many complexities such as:
Inserting new views,
New API for color mappings,
Difficult to test accessibility,
and etc.

mohammadcazig committed:
Adding a helper to merge two colors with different opacities.

mohammadcazig committed:
define nonnull for interface.

jverkoey reviewed (changes requested):
Please add at least a couple tests verifying black on white and white on black behavior with 100%, 25%, and 0% opacity.

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.m:
```diff
@@ -49,5 +49,19 @@ - (instancetype)initWithDefaults:(MDCColorSchemeDefaults)defaults {
+  CGFloat bRed = 0.0, bGreen = 0.0, bBlue = 0.0, bAlpha = 0.0;
+  [backgroundColor getRed:&bRed green:&bGreen blue:&bBlue alpha:&bAlpha];
+
+  NSAssert(bAlpha == 1, @"Background color with opacity is not supported.");
```
Please use a floating point equality check.
Why can't we support non-opaque background colors?
If this is a requirement, please note in the header.
In release code, this assert will be skipped, what is the fallback handling?

ianegordon commented on components/schemes/Color/src/MDCSemanticColorScheme.h:
```diff
@@ -116,4 +116,8 @@ typedef NS_ENUM(NSInteger, MDCColorSchemeDefaults) {
  */
 - (nonnull instancetype)initWithDefaults:(MDCColorSchemeDefaults)defaults;
 
+/** RGBA to RGB conversion */
```
Naming is hard.
blendColor:withBackgroundColor: may be more specific.
Please give a detailed description on how / why the two colors are merged.
(RGBA -> RGB is super detailed)

Is this the equivalent of blendOverlayMode?
https://developer.apple.com/documentation/coregraphics/cgblendmode/kcgblendmodeoverlay?language=objc

mohammadcazig replied:
I added documentation.
https://en.wikipedia.org/wiki/Alpha_compositing
it's different than blending images.

mohammadcazig committed:
tests for color merge functionality.

jverkoey commented on components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m:
```diff
@@ -67,4 +67,50 @@ - (void)testInitWithMaterialDefaults {
+
+}
+
+- (BOOL)compareColorsWithFloatPrecisionFirstColor:(UIColor *)firstColor
```
Add similar tests for blending on to black as well please.

mohammadcazig replied:
Done.

mohammadcazig committed:
Name change "merge" -> "blend"
full support for blending with any opacity.
More tests added.
Better Documentation.

jverkoey commented:
Failures appear to be legitimate:

```
ERROR: /Volumes/BuildData/tmpfs/src/github/repo/components/schemes/Color/BUILD:24:1: C++ compilation of rule '//components/schemes/Color:Color' failed (Exit 1)
components/schemes/Color/src/MDCSemanticColorScheme.m:58:17: error: incompatible pointer types sending 'double *' to parameter of type 'CGFloat * _Nullable' (aka 'float *') [-Werror,-Wincompatible-pointer-types]
  [color getRed:&red green:&green blue:&blue alpha:&alpha];
                ^~~~
```

ianegordon reviewed (dismissed):
Code LGTM, but please fix conversion errors and consider changing parameters from double to CGFloat to match the values.

romoore commented on components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m:
```diff
@@ -67,4 +67,58 @@ - (void)testInitWithMaterialDefaults {
+      [MDCSemanticColorScheme blendColor:blendColor withBackgroundColor:backgroundColor];
+  XCTAssertEqualObjects(resultColor, expectedColor);
+}
+- (BOOL)compareColorsWithFloatPrecisionFirstColor:(UIColor *)firstColor
```
Please add tests for HSB (including beyond 1.0), P3, and greyscale color spaces.

mohammadcazig replied:
Added more test for HSB and grey scale for now.

romoore replied:
Does this work for DisplayP3?

romoore commented on components/schemes/Color/src/MDCSemanticColorScheme.m:
```diff
@@ -22,6 +22,10 @@
                           blue:(CGFloat)((colorValue & 0xFF) / 255.0) alpha:1];
 }
 
+static double blendColorChannel(double value, double bValue, double alpha, double bAlpha) {
```
Please add documentation for parameters.

mohammadcazig replied:
Done.

romoore commented on components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m:
```diff
@@ -67,4 +67,58 @@ - (void)testInitWithMaterialDefaults {
+  UIColor *resultColor =
+      [MDCSemanticColorScheme blendColor:blendColor withBackgroundColor:backgroundColor];
+  UIColor *expectedColor =
+      [UIColor colorWithRed:240.0f/255.0f green:240.0f/255.0f blue:240.0f/255.0f alpha:1.0f];
```
Please add a test with non-uniform color channel information to detect if the color channels get mixed by accident.

mohammadcazig replied:
Done.

mohammadcazig committed:
ColorBlend:
Change the type to CGFloat from Double,
Updated Documentation and tests.

mohammadcazig committed:
Merge branch 'develop' of github.com:material-components/material-components-ios into Color_Merging_Helper

mohammadcazig committed:
fixing tests by putting expected values in proper precision.

romoore commented on components/schemes/Color/tests/unit/MDCSemanticColorSchemeTests.m:
```diff
@@ -149,10 +149,11 @@ - (BOOL)compareColorsWithFloatPrecisionFirstColor:(UIColor *)firstColor
-          roundf(fBlue) == roundf(sBlue) &&
-          roundf(fAlpha) == roundf(sAlpha));
+
+  return ((float)fRed == (float)sRed &&
```
Nit: Please consider using MDCFloatEquals

romoore reviewed (approved):
If not in this PR, please create a backlog issue to verify display P3 color space support.

mohammadcazig committed:
casting values to CGFloat so we support both 32bit and 64bits without losing precision.

mohammadcazig committed:
adding math to Build file.
